helper.py

```diff
@@ -1,41 +1,186 @@
+        # converts image to a base64 which can be sent as a long string to the browser.
+        img_b64 = im_2_b64(img).decode()
+        websockets.broadcast(
+            CONNECTIONS, "[CAMERA]" + img_b64
+        )  # sends image to all connected browsers.
+        print("Image broadcast.")
+
+    # politely stops watching file system.
+    picture_watcher.close()
+    loop.stop()
+    loop.close()
+
+
+async def wait_for_log_change():
+    loop = asyncio.get_event_loop()
+
+    bypass = False  # so first image is not ignored.
+    while not os.path.exists(log_input_file):
+        await asyncio.sleep(0.5)  # twiddle thumbs :)
+        if not bypass:
+            bypass = True
+    await log_watcher.setup(loop)
+    print("Log change watcher is running.")
+
+    while True:  # for all events
+        if not bypass:
+            event = await log_watcher.get_event()  # blocks until file changed
+        else:
+            bypass = False  # reset bypass
+
+        with open(log_input_file, "r") as l:
+            old_logs = l.read()
+        for c in range(10):
+            await asyncio.sleep(0.01)  # give it time to write the file.
+            try:  # this runs until the bot has finished writing the logs
+                with open(log_input_file, "r") as l:
+                    new_logs = l.read()
+                print("Opened logs successfully")
+                break
+            except:
+                print("Error opening logs: attempt \#" + str(c))
+
+        if c >= 9:
+            continue  # error with this file, go back and wait for next change.
+
+        new_logs.replace(old_logs, "")  # only new logs remain.
+        index = len(new_logs) - len(old_logs)
+        old_logs = new_logs
+        new_logs = new_logs[index:]
+
+        websockets.broadcast(CONNECTIONS, "[LOGS]" + new_logs)  # sends new logs.
+        print("Logs broadcast.")
+
+    # politely stops watching file system.
+    log_watcher.close()
+    loop.stop()
+    loop.close()
+
+
+async def register(websocket):  # Runs every time someone connects
+    CONNECTIONS.add(websocket)
+    print("Someone has connected to the websocket.")
+    for c in range(3):
+        time.sleep(0.1)  # give it time to write the file.
+        try:  # this runs until the bot has finished writing the image
+            img = Image.open(img_input_file)
+            img.load()
+            print("Opened successfully")
+            break
+        except:
+            print("Error opening file: attempt #" + str(c))
+    bypass = False
+    if c > 3:
+        bypass = True  # error with this file, go back and wait for next change.
+    if not bypass:
+        img = shrink_image(img)
+        img_b64 = im_2_b64(img).decode()
+        await websocket.send(img_b64)
+    try:
+        await websocket.wait_closed()
+    finally:
+        CONNECTIONS.remove(websocket)
+
+
+async def main():
+    async with serve(register, "0.0.0.0", 5001):
+        await asyncio.gather(
```

Comment:
Oooh this is nice. I didn't know about this.

Reply:
Took me ages to find it 🤣